### DISCUSSION TITLE
[Core Protocol] <Finalizing message sending from L2 to L1>

### DISCUSSION THREAD

**Discussion by kupermind** [ethereum-optimism/developers#357] [Core Protocol]
### Did you check the documentation?

- [X] Yes

### Did you read the specs?

- [X] Yes
- [ ] No

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

Trying to send messages from L2 to L1 via the [L2CrossDomainMessenger](https://optimistic.etherscan.io/address/0x4200000000000000000000000000000000000007) contract, wondering how finalize it on L1.

Here is a [description](https://docs.optimism.io/builders/app-developers/bridging/messaging#for-l2-to-l1-transactions-1) of the steps in order to send messages from L2 to L1.

Steps 2-3 from the description is not clear. Do I need to call functions like [proveWithdrawalTransaction()](https://github.com/ethereum-optimism/optimism/blob/65ec61dde94ffa93342728d324fecf474d228e1f/packages/contracts-bedrock/contracts/L1/OptimismPortal.sol#L242C14-L242C40) and / or [finalizeWithdrawalTransaction()](https://github.com/ethereum-optimism/optimism/blob/65ec61dde94ffa93342728d324fecf474d228e1f/packages/contracts-bedrock/contracts/L1/OptimismPortal.sol#L325) in [OptimismPortalProxy](https://etherscan.io/address/0xbEb5Fc579115071764c7423A4f12eDde41f106Ed)? Not sure how to get arguments for those function calls.

What about the `L2CrossDomainMessenger`'s [relayMessage()](https://github.com/ethereum-optimism/optimism/blob/65ec61dde94ffa93342728d324fecf474d228e1f/packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol#L303) function call that settles sent messages? Seems like the nonce parameter there is the one that fits considering that the [sendMessage()](https://github.com/ethereum-optimism/optimism/blob/65ec61dde94ffa93342728d324fecf474d228e1f/packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol#L259) function outputs the message nonce?

Thanks for the answers!

**Comment by 1195597989** [comment]
The sendMessage() function returns a message's nonce after sending the message on L2. This nonce is a unique number that identifies messages. When you call the L2CrossDomainMessenger relayMessage() function on L1, you need to provide this nonce as an argument.The relayMessage() function can find the correct message and pass it to the target contract.

**Reply by kupermind** [reply]
Thanks for the answer, there are actually more actions to do before the relay.

**Comment by kupermind** [comment]
Here is the description of how to act on L1 after the message is called on L2:
https://docs.optimism.io/builders/app-developers/tutorials/cross-dom-solidity#interact-with-the-l2-greeter

At the end, after the proof of message is done on L2, the actual contract and function that must be called on L1 is:
https://sepolia.etherscan.io/address/0x16Fc5058F25648194471939df75CF27A2fdC48BC#writeProxyContract
